### PR TITLE
Improved ISO15118-2 support

### PR DESCRIPTION
**Pull request by corneliusclaussen**
- Full support for ISO-2 AC charging with all init sequences
- Initial work on ISO-2 DC until PreCharge, no DIN SPEC support
- Basic SIL support for ISO
- Insulation monitoring device interface and SIL
- Adds optional Session logging (HTML and CSV)
- known limitations:
  * TCP session should be closed after ISO errors
  * parallel SLAC requests are not supported yet
  * AC SLAC retries not working
  * Session NMK is not regenerated for each session

Signed-off-by: Sebastian Lukas <sebastian.lukas@pionix.de>
Signed-off-by: Cornelius Claussen <cc@pionix.de>

### CODE REVIEW

File: config/test0.json [hunk 1]
```diff
@@ -0,0 +1,186 @@
+{
```
Comment (a-w50) [inline config/test0.json:1]: What are all these test*.json for, are they really necessary?

File: deploy.sh [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash -e
```
Comment (a-w50) [inline deploy.sh:1]: I would advise to move this deploy stuff and the gdb_test.sh somewhere else

File: interfaces/isolation_monitor.json [hunk 3]
```diff
@@ -0,0 +1,21 @@
+{
+    "description": "This interface defines an isolation monitoring device for DC charging",
```
Comment (a-w50) [inline interfaces/isolation_monitor.json:2]: what type of isolation? Maybe the description could be a bit more verbose here?

File: modules/EnergyManager/EnergyManager.hpp [hunk 4]
```diff
@@ -16,12 +16,11 @@
 // headers for required interface implementations
 #include <generated/energy/Interface.hpp>
 
+// ev@4bf81b14-a215-475c-a1d3-0a484ae48918:v1
```
Comment (a-w50) [inline modules/EnergyManager/EnergyManager.hpp:19]: why did this file got changed?

**Review by a-w50 — COMMENTED** [review]
That is quite a lot of stuff here.  I would definitely need more time to fully review it (which I would like to).
Some of the license headers are old (2021) or not using the year at all.

File: deploy.sh (hunk 2 above)
Comment (corneliusclaussen) [reply deploy.sh:1]: moved to deploy repo

File: modules/EnergyManager/EnergyManager.hpp (hunk 4 above)
Comment (corneliusclaussen) [reply modules/EnergyManager/EnergyManager.hpp:19]: Somebody accidentially removed the guarding tags earlier

File: modules/EvseManager/Charger.hpp [hunk 5]
```diff
@@ -228,6 +247,8 @@ class Charger {
 
     const float PWM_5_PERCENT = 0.05;
 
+    const int t_replug = 4000;
```
Comment (hikinggrass) [inline modules/EvseManager/Charger.hpp:250]: This should have a unit (ms in this case) in its name

File: modules/EvseManager/EvseManager.cpp [hunk 6]
```diff
@@ -29,13 +39,141 @@ void EvseManager::init() {
...
-        r_hlc[0]->call_set_rcd(false);
-        r_hlc[0]->call_set_max_current(0);
+    if (get_hlc_enabled()) {
+
+        // Set up EVSE ID
+        r_hlc[0]->call_set_EVSEID(config.evse_id);
+
+        // Set up auth options for HLC
+        Array payment_options;
+        if (config.payment_enable_eim)
```
Comment (hikinggrass) [inline modules/EvseManager/EvseManager.cpp:49]: We should always use {} in if blocks

File: modules/EvseManager/EvseManager.cpp [hunk 7]
```diff
@@ -29,13 +39,141 @@ void EvseManager::init() {
...
+        });
+
+        // Install debug V2G Messages handler if session logging is enabled
+        if (config.session_logging)
+            r_hlc[0]->subscribe_V2G_Messages([this](Object v2g_message) { log_v2g_message(v2g_message); });
+
+        // switch to DC mode for first session for AC with SoC
+        if (config.ac_with_soc) {
+
+            r_bsp->subscribe_event([this](std::string event) {
```
Comment (hikinggrass) [inline modules/EvseManager/EvseManager.cpp:155]: const std::string&

File: modules/EvseManager/EvseManager.cpp [hunk 8]
```diff
@@ -79,6 +266,10 @@ void EvseManager::ready() {
         // Inform charger about current charging current. This is used for slow OC detection.
         charger->setCurrentDrawnByVehicle(p["current_A"]["L1"], p["current_A"]["L2"], p["current_A"]["L3"]);
 
+        // Inform HLC about the power meter data
+        if (get_hlc_enabled())
```
Comment (hikinggrass) [inline modules/EvseManager/EvseManager.cpp:270]: if{}

File: modules/EvseManager/EvseManager.cpp [hunk 9]
```diff
@@ -306,5 +583,51 @@ bool EvseManager::reserved_for_different_token(const std::string& token) {
         return true;
     }
 }
+/*
```
Comment (hikinggrass) [inline modules/EvseManager/EvseManager.cpp:586]: large commented out code block, is this intentional?

File: modules/EvseManager/EvseManager.hpp [hunk 10]
```diff
@@ -109,8 +126,24 @@ class EvseManager : public Everest::ModuleBase {
     float local_max_current_limit;
     const float EVSE_ABSOLUTE_MAX_CURRENT = 80.0;
     bool slac_enabled;
+
+    std::mutex hlc_mutex;
+
     bool hlc_enabled;
 
+    /*
```
Comment (hikinggrass) [inline modules/EvseManager/EvseManager.hpp:134]: Ah I guess its intentional for implementing later?

File: modules/EvseManager/SessionLog.cpp [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+#include <date/tz.h>
+
+#include <boost/algorithm/string.hpp>
+#include <fmt/core.h>
+
+namespace module {
+
+SessionLog session_log;
+
+std::string to_rfc3339(std::chrono::time_point<date::utc_clock> t) {
```
Comment (hikinggrass) [inline modules/EvseManager/SessionLog.cpp:17]: everest-framework now has a Everest::Date::to_rfc3339 (and from_) function that you should use here

File: modules/EvseManager/SessionLog.cpp [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+}
+
+SessionLog::SessionLog() {
+    session_active = false;
+    enabled = false;
+    xmloutput = true;
+}
+
+SessionLog::~SessionLog() {
+    if (logfile_csv.is_open())
```
Comment (hikinggrass) [inline modules/EvseManager/SessionLog.cpp:28]: if{}

File: modules/EvseManager/SessionLog.cpp [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+    logpath = path;
+}
+
+void SessionLog::enable() {
+    enabled = true;
+}
+
+void SessionLog::startSession(const std::string& session_id) {
+    if (enabled) {
+        if (session_active)
```
Comment (hikinggrass) [inline modules/EvseManager/SessionLog.cpp:42]: if{}

File: modules/EvseManager/evse/evse_managerImpl.cpp [hunk 14]
```diff
@@ -133,7 +148,14 @@ void evse_managerImpl::ready() {
             if (p.contains("energy_Wh_export") && p["energy_Wh_export"].contains("total"))
```
Comment (hikinggrass) [inline modules/EvseManager/evse/evse_managerImpl.cpp:148]: if{}

File: modules/EvseManager/evse/evse_managerImpl.cpp [hunk 15]
```diff
@@ -133,7 +148,14 @@ void evse_managerImpl::ready() {
...
                 se["session_finished"]["energy_Wh_export"] = p["energy_Wh_export"]["total"];
 
+            double energy_import = 0.;
+            if (se["session_finished"].contains("energy_Wh_import"))
```
Comment (hikinggrass) [inline modules/EvseManager/evse/evse_managerImpl.cpp:152]: if{}

File: modules/EvseManager/evse/evse_managerImpl.cpp [hunk 16]
```diff
@@ -217,6 +245,8 @@ bool evse_managerImpl::handle_cancel_charging(std::string& reason) {
         std::lock_guard<std::mutex> lock(session_mutex);
         cancel_session_reason = reason;
     }
+    if (mod->get_hlc_enabled())
```
Comment (hikinggrass) [inline modules/EvseManager/evse/evse_managerImpl.cpp:248]: if{}

File: modules/EvseManager/Charger.cpp [hunk 17]
```diff
@@ -299,71 +331,83 @@ void Charger::runStateMachine() {
...
+                signalEvent(EvseEvent::ChargingPausedEV);
+                r_bsp->call_allow_power_on(true);
+                // make sure we are enabling PWM
+                if (hlc_use_5percent_current_session)
+                    update_pwm_now(PWM_5_PERCENT);
+                else
+                    update_pwm_now(ampereToDutyCycle(getMaxCurrent()));
+            } else {
+                // update PWM if it has changed and 5 seconds have passed since last update
+                if (!hlc_use_5percent_current_session)
```
Comment (hikinggrass) [inline modules/EvseManager/Charger.cpp:401]: if{}

File: modules/EvseManager/evse/evse_managerImpl.cpp [hunk 18]
```diff
@@ -150,6 +172,12 @@ void evse_managerImpl::ready() {
                 std::lock_guard<std::mutex> lock(session_mutex);
                 se["session_cancelled"]["reason"] = cancel_session_reason;
             }
+
+            double energy_import = 0.;
+            if (se["session_cancelled"].contains("energy_Wh_import"))
```
Comment (hikinggrass) [inline modules/EvseManager/evse/evse_managerImpl.cpp:177]: if{}

File: modules/JsCarV2G/index.js [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+
+    mqtt.subscribe(mqtt_paths.vars, (mod, raw_data) => {
+        if (mod === undefined) {
+        evlog.warning('Already received data, but framework is not ready yet');
+        return;
+        }
+        evlog.debug(`raw data: ${raw_data}`);
+        const data = JSON.parse(raw_data);
+        evlog.debug(`stringified data: ${data}`);
+        if (data.impl_id === 'ev') mod.provides.ev.publish[data.var](data.val);
```
Comment (hikinggrass) [inline modules/JsCarV2G/index.js:91]: if{}

File: modules/JsIMDSimulator/index.js [hunk 20]
```diff
@@ -0,0 +1,27 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2020 - 2021 Pionix GmbH and Contributors to EVerest
```
Comment (hikinggrass) [inline modules/JsIMDSimulator/index.js:2]: 2022

File: modules/EvseManager/EvseManager.cpp (hunk 9 above)
Comment (corneliusclaussen) [reply modules/EvseManager/EvseManager.cpp:595]: will remove

File: modules/EvseManager/EvseManager.hpp (hunk 10 above)
Comment (corneliusclaussen) [reply modules/EvseManager/EvseManager.hpp]: removed

File: modules/EvseManager/Charger.hpp (hunk 5 above)
Comment (corneliusclaussen) [reply modules/EvseManager/Charger.hpp]: done

**Review by hikinggrass — APPROVED** [review]

**Pull request merged**